### PR TITLE
rgw: throw error when more than one prefix in RGWPutLC::execute()

### PR DESCRIPTION
**Pull request by Huber-ming**
backport tracker:  https://tracker.ceph.com/issues/46702
Signed-off-by: Shengming Zhang <zhangsm01@inspur.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_lc_s3.cc [hunk 1]
```diff
@@ -139,9 +139,20 @@ void LCFilter_S3::decode_xml(XMLObj *obj)
     single_cond = true;
   }
 
-  RGWXMLDecoder::decode_xml("Prefix", prefix, o);
-  if (!prefix.empty())
+  auto prefix_iter = o->find("Prefix");
+  std::string temp_prefix;
+  std::vector<std::string> prefixs;
+  while (auto prefix_xml = prefix_iter.get_next()){
+    decode_xml_obj(temp_prefix, prefix_xml);
```
Comment (dang) [inline src/rgw/rgw_lc_s3.cc:146]: Decoding and storing all the prefixes is overkill here, if we're just going to error out.  Just do a single get_next() and decode, then do another get_next(), and if it returns something, throw the error.

Comment (Huber-ming) [reply src/rgw/rgw_lc_s3.cc]: thanks, I reset and re-push a new commit.

**Comment by Huber-ming** [comment]
jenkins retest this please

**Comment by Huber-ming** [comment]
jenkins retest this please

**Comment by Huber-ming** [comment]
jenkins retest this please

File: src/rgw/rgw_lc_s3.cc [hunk 2]
```diff
@@ -139,7 +139,16 @@ void LCFilter_S3::decode_xml(XMLObj *obj)
...
   }
 
-  RGWXMLDecoder::decode_xml("Prefix", prefix, o);
+  {
+    auto prefix_iter = o->find("Prefix");
+    auto prefix_val = prefix_iter.get_next();
+    if(prefix_val) {
+      decode_xml_obj(prefix, prefix_val);
+      if(prefix_iter.get_next()){
+        throw RGWXMLDecoder::err("Bad filter: the number of prefix is more than one");
```
Comment (ivancich) [inline src/rgw/rgw_lc_s3.cc:148]: For proper English, I would say: "the number of prefixes is more than one".

For more natural English, I would say: "there is more than one prefix".
Or even better: "there should not be more than one prefix".

**Review by ivancich — APPROVED** [review]
Approved, although with wording suggestions.

**Comment by mattbenjamin** [comment]
@dang do you remember why this change was needed?

File: src/rgw/rgw_lc_s3.cc [hunk 3]
```diff
@@ -139,7 +139,16 @@ void LCFilter_S3::decode_xml(XMLObj *obj)
     single_cond = true;
   }
 
-  RGWXMLDecoder::decode_xml("Prefix", prefix, o);
+  {
+    auto prefix_iter = o->find("Prefix");
+    auto prefix_val = prefix_iter.get_next();
+    if(prefix_val) {
```
Comment (theanalyst) [inline src/rgw/rgw_lc_s3.cc:145]: probably invalid if prefix_iter is invalid, should be `prefix_iter && prefix_val`

**Pull request closed without merging**